### PR TITLE
support vterm-copy-mode

### PR DESCRIPTION
**Pull request by jixiuf**
#75

### CODE REVIEW

File: vterm.el [hunk 1]
```diff
@@ -276,6 +282,19 @@ If nil, never delay")
...
+(define-minor-mode vterm-copy-mode
+  "Toggle vterm copy mode."
+  :group 'vterm
+  :lighter " VtermCopy"
+  (if vterm-copy-mode
+      (progn                            ;enable vterm-copy-mode
+        (use-local-map vterm-copy-map)
+        (vterm-send-key "s" nil nil t))
+    (use-local-map vterm-raw-map)
+    (vterm-send-key "q" nil nil t)))
```
Comment (akermu) [inline vterm.el]: For this to work you have to add `ixon` to the stty call after the sane here: https://github.com/akermu/emacs-libvterm/blob/ee07b5a1e798a8b45c242de2bcdabc009ff055be/vterm.el#L235

File: vterm.el [hunk 2]
```diff
@@ -206,6 +206,8 @@ and background color are used for bright color. "
   "Delay for refreshing the buffer after receiving updates from libvterm.
 Improves performance when receiving large bursts of data.
 If nil, never delay")
+(defvar vterm-copy-map  (make-sparse-keymap))
+(defvar vterm-raw-map  (make-sparse-keymap))
```
Comment (akermu) [inline vterm.el]: Please remove those two maps and instead use `(use-local-keymap nil)` to disable `vterm-mode-map` and use (use-local-keymap vterm-mode-map)` to activate it again.

File: vterm.el [hunk 3]
```diff
@@ -276,6 +282,19 @@ If nil, never delay")
                                        unless (member key vterm-keymap-exceptions)
                                        collect key))))
 
+
+(define-minor-mode vterm-copy-mode
+  "Toggle vterm copy mode."
+  :group 'vterm
+  :lighter " VtermCopy"
```
Comment (akermu) [inline vterm.el]: Maybe change this to `" VTermCopy"`.

File: vterm.el (hunk 1 above)
Comment (jixiuf) [reply vterm.el]: done

File: vterm.el (hunk 2 above)
Comment (jixiuf) [reply vterm.el]: I removed  `term-raw-map` and leave `vterm-copy-map` as the keymap for `vterm-mode`

File: vterm.el (hunk 3 above)
Comment (jixiuf) [reply vterm.el]: done

**Comment by Sbozzolo** [comment]
I checked out this PR. With `zsh`, `C-c C-t` triggers forward search.

**Comment by jixiuf** [comment]
> I checked out this PR. With `zsh`, `C-c C-t` triggers forward search.

can you paste the result of `C-hkC-cC-t` like this 
```
C-c C-t runs the command vterm-copy-mode (found in vterm-mode-map), which is an interactive Lisp
closure in ‘vterm.el’.

It is bound to C-c C-t.

(vterm-copy-mode &optional ARG)

Toggle vterm copy mode.

If called interactively, enable Vterm-Copy mode if ARG is positive, and
disable it if ARG is zero or negative.  If called from Lisp,
also enable the mode if ARG is omitted or nil, and toggle it
if ARG is ‘toggle’; disable the mode otherwise.

[back]
```

and all your keybinding about `C-cC-t`
and I try bind `C-cC-t` to isearch-forward and the bug you mention doesn't reproduced
```
(global-set-key (kbd "C-c C-t") 'isearch-forward)
```
and tell me how can I reproduce .

**Comment by Sbozzolo** [comment]
I removed most of my customizations to test this patch. This is the output of `helpful` regarding `C-c C-t`:

```
vterm-copy-mode is an interactive function defined in vterm.el.

Signature
(vterm-copy-mode &optional ARG)

Documentation
Toggle vterm copy mode.

Key Bindings
vterm-copy-map C-c C-t
vterm-mode-map C-c C-t

References
References in vterm.el:
(define-key vterm-mode-map ...) 1 reference
(define-key vterm-copy-map ...) 1 reference

Find all references Find callees

Debugging
Enable edebug Enable tracing
Disassemble Forget

Source Code
;; Defined in ~/.emacs.d/lisp/emacs-libvterm/vterm.el
(define-minor-mode vterm-copy-mode
  "Toggle vterm copy mode."
  :group 'vterm
  :lighter " VTermCopy"
  :keymap vterm-copy-map
  (if vterm-copy-mode
      (progn                            ;enable vterm-copy-mode
        (use-local-map nil)
        (vterm-send-key "s" nil nil t))
    (use-local-map vterm-mode-map)
    (vterm-send-key "q" nil nil t)))

Symbol Properties
event-symbol-element-mask
  (vterm-copy-mode 0)
event-symbol-elements
  (vterm-copy-mode)
modifier-cache
  ((0 . vterm-copy-mode))
variable-documentation
  "Non-nil if Vterm-Copy mode is enabled.\nUse the command `vterm-copy-mode' to change this variable."
```

The `search forward` I mention is the shell one, not Emacs'.
This is what I see when I enter the copy mode:

**Comment by jixiuf** [comment]
you remove `C-c` from vterm-keymap-exceptions?

**Comment by Sbozzolo** [comment]
This is my configuration:

``` emacs-lisp
(use-package vterm
  :load-path  "~/.emacs.d/lisp/emacs-libvterm/"
  :defer t
  :commands (vterm)
  :config
  (add-hook 'vterm-exit-functions #'(lambda (buf)
                                      (when buf (kill-buffer buf))))
  ;; Remove binding of <f#>, and move it to S-<f#>
  (cl-loop for number from 1 to 12 do
           (progn
             (define-key vterm-mode-map (kbd (format "<f%i>" number)) nil)
             )
           )

  ;; I cannot get the loop to work...
  (define-key vterm-mode-map (kbd "s-<f0>") '(lambda () (interactive) (vterm-send-key "<f0>")))
  (define-key vterm-mode-map (kbd "s-<f1>") '(lambda () (interactive) (vterm-send-key "<f1>")))
  (define-key vterm-mode-map (kbd "s-<f2>") '(lambda () (interactive) (vterm-send-key "<f2>")))
  (define-key vterm-mode-map (kbd "s-<f3>") '(lambda () (interactive) (vterm-send-key "<f3>")))
  (define-key vterm-mode-map (kbd "s-<f4>") '(lambda () (interactive) (vterm-send-key "<f4>")))
  (define-key vterm-mode-map (kbd "s-<f5>") '(lambda () (interactive) (vterm-send-key "<f5>")))
  (define-key vterm-mode-map (kbd "s-<f6>") '(lambda () (interactive) (vterm-send-key "<f6>")))
  (define-key vterm-mode-map (kbd "s-<f7>") '(lambda () (interactive) (vterm-send-key "<f7>")))
  (define-key vterm-mode-map (kbd "s-<f8>") '(lambda () (interactive) (vterm-send-key "<f8>")))
  (define-key vterm-mode-map (kbd "s-<f9>") '(lambda () (interactive) (vterm-send-key "<f9>")))
  (define-key vterm-mode-map (kbd "s-<f10>") '(lambda () (interactive) (vterm-send-key "<f10>")))
  (define-key vterm-mode-map (kbd "s-<f11>") '(lambda () (interactive) (vterm-send-key "<f11>")))
  (define-key vterm-mode-map (kbd "s-<f12>") '(lambda () (interactive) (vterm-send-key "<f12>")))

  (defvar vterm-user "")
  (make-variable-buffer-local 'vterm-user)
  (defvar vterm-host "")
  (make-variable-buffer-local 'vterm-host)
  (defvar vterm-pwd "")
  (make-variable-buffer-local 'vterm-pwd)
  (defvar vterm-cmd "")
  (make-variable-buffer-local 'vterm-cmd)

  (defun vterm-set-title-hook (title) ;title = user@host@lastcmd:path  or user@host:path
    (let* ((tokens (split-string title ":" ))
           dir)
      (when (equal 2 (length tokens))
        (setq vterm-pwd (string-trim-left (nth 1 tokens)))
        (setq tokens (split-string (nth 0 tokens) "@" ))
        (when (>  (length tokens) 1)
          (setq vterm-user (nth 0 tokens))
          (setq vterm-host (nth 1 tokens))
          (when (and (nth 2 tokens)
                     (not (string-empty-p (or (nth 2 tokens) ""))))
            (setq vterm-cmd (nth 2 tokens))))
        (setq dir
              (file-name-as-directory
               (if (and (string= vterm-host (system-name))
                        (string= vterm-user (user-real-login-name)))
                   (expand-file-name vterm-pwd)
                 (concat "/-:" vterm-user "@" vterm-host ":"
                         vterm-pwd))))
        (when (ignore-errors (file-directory-p dir))
          (cd-absolute dir))
        (message title)
        ;; (message "pwd=%s,user=%s,host=%s,cmd=%s d=%s"
        ;;          vterm-pwd vterm-user vterm-host vterm-cmd dir)
        (rename-buffer (format "term %s %s@%s:%s" vterm-cmd vterm-user vterm-host vterm-pwd ) t)
        )))

  (add-hook 'vterm-set-title-functions 'vterm-set-title-hook)
  )

(use-package vterm-toggle
  :load-path  "~/.emacs.d/lisp/vterm-toggle/"
  :defer t
  :commands (vterm-toggle-cd)
  )
```

**Comment by jixiuf** [comment]
still can't reproduce with your config 
can you reproduce by  running `make run`  in your `~/.emacs.d/lisp/emacs-libvterm/build`
and make sure you are using the last version

**Comment by Sbozzolo** [comment]
Same problem


Could it be that I am using 'oh-my-zsh'?

**Comment by jixiuf** [comment]
yeah ,reproduced with on-my-zsh 
trigged by `ctrl-s` 
```
(vterm-send-key "s" nil nil t)
```
```
bindkey '^s' history-incremental-search-forward
```
I will try to find out the reason

File: vterm.el [hunk 4]
```diff
@@ -276,6 +256,49 @@ If nil, never delay")
                                        unless (member key vterm-keymap-exceptions)
                                        collect key))))
 
+;; Keybindings
+(define-key vterm-mode-map [tab]                       #'vterm-send-tab)
+(define-key vterm-mode-map (kbd "TAB")                 #'vterm-send-tab)
+(define-key vterm-mode-map [backtab]                   #'vterm--self-insert)
+(define-key vterm-mode-map [backspace]                 #'vterm-send-backspace)
```
Comment (akermu) [inline vterm.el]: Why did you change these? For terminal support?

Comment (jixiuf) [reply vterm.el]: yes
